File: Exercise_1.cs
class Stack { 
    // Did this code successfully run on Leetcode :
    // Any problem you faced while coding this :
        static int MAX = 1000; 
        int top; 
        int[] a = new int[MAX]; // Maximum size of Stack 
      
        // Time Complexity : O(1)
        // Space Complexity : O(1)
        public bool IsEmpty() 
        { 
            //Write your code here 
            return top == -1;
        } 
    
        public Stack() 
        { 
            //Initialize your constructor 
            top = -1;
        } 
      
        // Time Complexity : O(1)
        // Space Complexity : O(1)
        public bool Push(int x) 
        { 
            //Check for stack Overflow
            //Write your code here
            if(++top < MAX){
                a[top] = x;
                return true;
            } else{
                top--;
                System.Console.WriteLine("Stack Overflow");
                return false;
            }
        } 
      
        // Time Complexity : O(1)
        // Space Complexity : O(1)
        public int Pop() 
        { 
            //If empty return 0 and print " Stack Underflow"
            //Write your code here
            if(top<0)
            {
                System.Console.WriteLine("Stack Underflow");
                return 0;
            } else {
                int temp = a[top];
                top --;
                return temp;
            }
        } 
        
        // Time Complexity : O(1)
        // Space Complexity : O(1)
        public int Peek() 
        { 
            //Write your code here
            return IsEmpty() ? 0 : a[top];
        } 
    } 
      
    // Driver code 
    class Main { 
        public static void main(String[] args) 
        { 
            Stack s = new Stack(); 
            s.Push(10); 
            s.Push(20); 
            s.Push(30); 
            System.Console.WriteLine(s.Pop() + " Popped from stack"); 
        } 
    }

File: Exercise1.go
package main

import (
	"fmt"
)

type Stack struct {
  items []int
}

func NewStack() *Stack {
  return &Stack{
    items: []int{},
  }
}

func (s *Stack) isEmpty() bool  {
  return len(s.items) == 0
}

func (s *Stack) Push(x int) {
  s.items = append(s.items, x)
}

func (s *Stack) Pop() int {
  if s.isEmpty() {
    panic("Stack is empty, nothing to pop....")
  }
  out := s.items[len(s.items)-1]
  s.items = s.items[:len(s.items)-1]
  return out
}

func (s *Stack) Peek() int {
  if s.isEmpty() {
    panic("Stack is empty, nothing to pop....")
  }
  return s.items[len(s.items)-1]
}



func main() {
	s := NewStack()
  s.Push(10)
  fmt.Println("Peek: ", s.Peek())
  fmt.Println("isEmpty: ", s.isEmpty())

  top := s.Pop()
  fmt.Println("Top: ", top)
  fmt.Println("isEmpty: ", s.isEmpty())
  s.Peek()
}
